transactor/src/main/scala/io/mediachain/transactor/Copycat.scala

```diff
@@ -0,0 +1,186 @@
+    ref: Reference
+  ) extends Query[Option[Reference]]
+  
+  case class JournalCommitEvent(entry: JournalEntry)
+
+  class JournalStateMachine(
+    val datastore: Datastore
+  ) extends StateMachine with Snapshottable with SessionListener {
+    sealed abstract class CanonicalReference {
+      def chain: Option[Reference]
+    }
+    case class EntityReference(chain: Option[Reference]) extends CanonicalReference
+    case class ArtefactReference(chain: Option[Reference]) extends CanonicalReference
+    
+    private var seqno: BigInt = 0
+    private var index: MMap[Reference, CanonicalReference] = new MHashMap    
+    private var clients: MSet[ServerSession] = new MHashSet // this wanted to be called sessions
+
+    
+    private def CommitError(what: String) = Xor.left(JournalCommitError(what))
+    
+    // Journal Interface
+    def insert(commit: Commit[JournalInsert]): Xor[JournalError, CanonicalEntry] = {
+      try {
+        insertRecord(commit.operation)
+      } finally {
+        commit.release()
+      }
+    }
+    
+    private def insertRecord(cmd: JournalInsert): Xor[JournalError, CanonicalEntry] = {
+      val rec = cmd.record
+      val ref = datastore.put(rec)
+      index.get(ref) match {
+        case Some(_) => CommitError("duplicate insert")
+        case None => {
+          rec match {
+            case Entity(_) => {
+              index += (ref -> EntityReference(None))
+            }
+            case Artefact(_) => {
+              index += (ref -> ArtefactReference(None))
+            }
+          }
+          
+          val entry = CanonicalEntry(nextSeqno(), ref)
+          publishCommit(entry)
+          Xor.right(entry)
+        }
+      }
+    }
+
+    def update(commit: Commit[JournalUpdate]): Xor[JournalError, ChainEntry] = {
+      try {
+        updateChain(commit.operation)
+      } finally {
+        commit.release()
+      }
+    }
+    
+    private def updateChain(cmd: JournalUpdate): Xor[JournalError, ChainEntry] = {
+      val ref = cmd.ref
+      val cell = cmd.cell
+      index.get(ref) match {
+        case None => CommitError("invalid reference")
+        case Some(cref) => {
+          cref match {
+            case EntityReference(chain) => {
+              cell match {
+                case EntityChainCell(entity, xchain, meta) => {
+                  if (checkUpdate(ref, chain, entity, xchain)) {
+                    val newcell = EntityChainCell(ref, chain, meta)
+                    val newchain = datastore.put(newcell)
+                    index.put(ref, EntityReference(Some(newchain)))
+                    
+                    val entry = ChainEntry(nextSeqno(), ref, newchain, chain)
+                    publishCommit(entry)
+                    Xor.right(entry)
+                  } else CommitError("invalid chain cell")
+                }
+                case _ => CommitError("invalid chain")
+              }
+            }
+            case ArtefactReference(chain) => {
```

Comment:
Here

```
case (ArtefactReference(chain), ArtefactChainCell(...)) => ...
case (ArtefactReference(_), _) => 
```

would immediately tell us why the error occurred (mismatched reference and cell)